tokio/src/io/util/mem.rs

```diff
@@ -0,0 +1,178 @@
+    is_closed: bool,
+    /// If the `read` side has been polled and is pending, this is the waker
+    /// for that parked task.
+    read_waker: Option<Waker>,
+}
+
+// ===== impl MemStream =====
+
+impl MemStream {
+    /// Create a new pair of `MemStream`s that act like a pair of connected sockets.
+    pub fn pair() -> (MemStream, MemStream) {
+        let one = Arc::new(Mutex::new(Pipe::new()));
+        let two = Arc::new(Mutex::new(Pipe::new()));
+
+        (
+            MemStream {
+                read: one.clone(),
+                write: two.clone(),
+            },
+            MemStream {
+                read: two,
+                write: one,
+            },
+        )
+    }
+}
+
+impl AsyncRead for MemStream {
+    fn poll_read(self: Pin<&mut Self>, cx: &mut task::Context<'_>, buf: &mut [u8])
+        -> Poll<std::io::Result<usize>>
+    {
+        Pin::new(&mut *self.read.lock().unwrap()).poll_read(cx, buf)
+    }
+}
+
+impl AsyncWrite for MemStream {
+    fn poll_write(self: Pin<&mut Self>, cx: &mut task::Context<'_>, buf: &[u8])
+        -> Poll<std::io::Result<usize>>
+    {
+        Pin::new(&mut *self.write.lock().unwrap()).poll_write(cx, buf)
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut task::Context<'_>)
+        -> Poll<std::io::Result<()>>
+    {
+        Pin::new(&mut *self.write.lock().unwrap()).poll_flush(cx)
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut task::Context<'_>)
+        -> Poll<std::io::Result<()>>
+    {
+        Pin::new(&mut *self.write.lock().unwrap()).poll_shutdown(cx)
+    }
+}
+
+impl Drop for MemStream {
+    fn drop(&mut self) {
+        // notify the other side of the closure
+        self.write.lock().unwrap().close();
+    }
+}
+
+// ===== impl Pipe =====
+
+impl Pipe {
+    fn new() -> Self {
+        Pipe {
+            buffer: BytesMut::new(),
+            is_closed: false,
+            read_waker: None,
+        }
+    }
+
+    fn close(&mut self) {
+        self.is_closed = true;
+        if let Some(waker) = self.read_waker.take() {
+            waker.wake();
+        }
+    }
+}
+
+impl AsyncRead for Pipe {
+    fn poll_read(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>, buf: &mut [u8])
+        -> Poll<std::io::Result<usize>>
+    {
+        if self.buffer.has_remaining() {
+            let max = self.buffer.remaining().min(buf.len());
+            self.buffer.copy_to_slice(&mut buf[..max]);
+            Poll::Ready(Ok(max))
+        } else if self.is_closed {
+            Poll::Ready(Ok(0))
+        } else {
+            self.read_waker = Some(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+}
+
+impl AsyncWrite for Pipe {
+    fn poll_write(mut self: Pin<&mut Self>, _: &mut task::Context<'_>, buf: &[u8])
+        -> Poll<std::io::Result<usize>>
+    {
+        if self.is_closed {
+            return Poll::Ready(Err(std::io::ErrorKind::BrokenPipe.into()));
+        }
+        self.buffer.extend_from_slice(buf);
```

Comment:
This should also have backpressure to make it a reasonable simulation of other sockets.
Either opt for a default buffer size or allow the user to configure the internal buffer size

Reply:
I did want to wait on defining an API for specifying a maximum, but perhaps including a default maximum should be considered, since without one, users may assume there's no backpressure and then be broken if we add a default.

Then again, maybe there shouldn't be a default unless you specifically set one...

Reply:
I now saw that @bdonlan commented the same.

The issue I have without limit is that the semantics are vastly different than from a real socket - which will always have some kind of backpressure. If you use this for a unit-test or even for communication between components (which is a super legit use-case) you could run into OOM when there is no reader attached.

I think a default limit is a good start. The problem with that is that if tests start to depend on the default limit they might fail if you change it later on. E.g. some tests might assume that an initial write of 10kB will never block. If we then later on decrease the limit it would block, and only half of the data would be written. Been there with my own implementation of this.

So if we pick a value we should stick with it. Something between 64kB and 512kB might be a decent value.

Allowing users to configure the buffer size however also shouldn't be that hard. It's just a parameter like the channel size is a parameter on those. I don't expect this to be require a lot of additional parameters in the future, so a builder might not be worth it.

Reply:
Perhaps just having a `pub fn stream(limit: usize) -> (MemStream, MemStream)` and `pub fn unbounded_stream() -> (MemStream, MemStream)` would be best, like with the channel API?